src/utils/token_to_fiat_oracle.ts

```diff
@@ -0,0 +1,137 @@
+ * 1. Import the CoinGecko wrapper.
+ * 2. Define a CoinGecko client as private readonly at the
+ *   constructor arguments. So we can test it.
+ * 3. On each call to the API, replace the fetch for the actual
+ *   API call.
+ *
+ * Examples
+ *
+ * 1. import CoinGecko from 'coingecko-api'
+ * 2. constructor(private readonly client = new CoinGecko()) {}
+ * 3. const response = await this.client.ping() // for the ping() function
+ *    const response = await this.client.simple.price(params) // for getting the price of a currency
+ */
+
+export class TokenToFiatOracle implements FiatOracle {
+	private cachedPrices: TokenFiatRate[] = [];
+	private cacheTimestamp = 0;
+	private syncPromise?: Promise<void>;
+
+	/**
+	 * @param {FiatID[]} fiats The currency IDs to quote tokens in
+	 * @param {TokenID[]} tokens The token ID of the requested coin
+	 * @param {number} cacheLifespan Milliseconds that has to pass until the cache is invalid
+	 * @param {@link fetch} _fetch The standard fetch method. Defined here for testing propuses
+	 * @return {TokenToFiatOracle} A class representing the oracle for the selected currencies
+	 */
+	constructor(
+		private readonly fiats: FiatID[] = ['USD'],
+		private readonly tokens: TokenID[] = ['ARWEAVE'],
+		private readonly cacheLifespan = pollingIntervalMilliseconds,
+		private readonly _fetch = fetch
+	) {}
+
+	private get currentlyFetchingPrice(): boolean {
+		return !!this.syncPromise;
+	}
+
+	private get tokensCSV(): string {
+		return this.tokens.map(tokenIdToThirdParty).join(',');
+	}
+
+	private get fiatCSV(): string {
+		return this.fiats.map(fiatIdToThirdParty).join(',');
+	}
+
+	private get shouldRefreshCacheData(): boolean {
+		const currentTimestamp = Date.now();
+		const deltaTimestamp = currentTimestamp - this.cacheTimestamp;
+		return deltaTimestamp >= this.cacheLifespan;
+	}
+
+	/**
+	 * @param {FiatID} fiat The fiat currency ID
+	 * @param {TokenID} token The token ID
+	 * @throws {@link Error} If no such pair cached
+	 * @returns {Promise<TokenToFiatPrice>} The cached price value
+	 */
+	public async getPriceForFiatTokenPair(pair: TokenFiatPair): Promise<TokenFiatRate> {
+		await this.checkCache();
+		const pricePair = this.cachedPrices.find(cachePairFilterFactory(pair));
+		if (!pricePair) {
+			throw new Error(`No such pair (${pair.fiat}, ${pair.token})`);
+		}
+		return pricePair;
+	}
+
+	private async checkCache(): Promise<void> {
+		if (this.currentlyFetchingPrice) {
+			return this.syncPromise;
+		}
+		if (this.shouldRefreshCacheData) {
+			this.syncPromise = this.fetchPrices();
+			await this.syncPromise;
+			delete this.syncPromise;
+		}
+	}
+
+	private async fetchPrices(): Promise<void> {
+		const queryUrl = this.getQueryRequestUrl();
+		const fetchResponse = await this._fetch(queryUrl);
+		const responseData: CoinGeckoPriceResponseData = await fetchResponse.json();
+		this.fiats.forEach((fiat) => {
+			this.tokens.forEach((token) => {
+				const priceValue = responseData[token][fiat];
```

Comment:
Next, there is an error here when grabbing the priceValue. Looks like its caused by the capitalization change, the CoinGecko response is always uncapitalized:

```
Failed to use fiat oracle: TypeError: Cannot read property 'USD' of undefined
    at token_to_fiat_oracle.ts:105
    at Array.forEach (<anonymous>)
    at token_to_fiat_oracle.ts:102
    at Array.forEach (<anonymous>)
    at TokenToFiatOracle.fetchPrices (token_to_fiat_oracle.ts:101)
```

Here this might be helpful, swap `App.tsx` component into this temporarily to self-test your entry point on refresh:

```ts
export default function App(): JSX.Element {
	useEffect(() => {
		async function getFiatToken() {
			try {
				const fiatOracle = new TokenToFiatOracle();

				const tokenPair = await fiatOracle.getPriceForFiatTokenPair({
					fiat: 'USD',
					token: 'ARWEAVE'
				});
				console.log('token pair: ', tokenPair);
			} catch (error) {
				console.log('Failed to use fiat oracle:', error);
			}
		}

		getFiatToken();
	}, []);
	
	return (
		<AppContainer>
			<Calculator></Calculator>
			<Faq></Faq>
		</AppContainer>
	);
}
```

Reply:
There was a conflict on the token and fiat IDs. Now there are types and functions for mapping them introduced by this commit 77f77c77d7cb946917b6232d0fb1ad6813a6a395

=====================

Thanks for the self-testing code example!